### ISSUE TITLE
TypeError: argument of type 'method' is not iterable on install of mmdetection

### ISSUE THREAD

**Issue by pdmct** [open-mmlab/mim#203]
Thanks for your error report and we appreciate it a lot.

**Checklist**

1. I have searched related issues but cannot get the expected help. - done
2. I have read the [FAQ documentation](https://openmim.readthedocs.io/en/latest/faq.html) but cannot get the expected help.
 -- there is no FAQ page at this URL 
4. The bug has not been fixed in the latest version. - not as far as I can tell

**Describe the bug**
I am following the installation instructions for mmyolo on: [https://github.com/open-mmlab/mmyolo/blob/main/docs/en/tutorials/custom_installation.md](https://github.com/open-mmlab/mmyolo/blob/main/docs/en/tutorials/custom_installation.md)

mim throws an error: `TypeError: argument of type 'method' is not iterable`

I had previously installled the following components successfully:
```
pip3 install openmim
mim install "mmengine>=0.6.0"
mim install "mmcv>=2.0.0rc4,<2.1.0"
```

**Reproduction**

1. What command or script did you run?

```none
mim install "mmdet>=3.0.0rc6,<3.1.0"
```

2. Did you make any modifications on the code or config? Did you understand what you have modified?
3. What dataset did you use?

**Environment**

1. Please run `python mmdet/utils/collect_env.py` to collect necessary environment information and paste it here.
-- can't run this as mmdet fails to install
```
mim --version
mim, version 0.3.6

mmyolo git:(main) ✗ pip freeze
addict==2.4.0
certifi==2022.12.7
charset-normalizer==3.1.0
click==8.1.3
colorama==0.4.6
contourpy==1.0.7
cycler==0.11.0
filelock==3.10.0
fonttools==4.39.0
idna==3.4
Jinja2==3.1.2
kiwisolver==1.4.4
Markdown==3.4.1
markdown-it-py==2.2.0
MarkupSafe==2.1.2
matplotlib==3.7.1
mdurl==0.1.2
mmcv==2.0.0rc4
mmengine==0.6.0
model-index==0.1.11
mpmath==1.3.0
networkx==3.0
numpy==1.24.2
opencv-python==4.7.0.72
openmim==0.3.6
ordered-set==4.1.0
packaging==23.0
pandas==1.5.3
Pillow==9.4.0
Pygments==2.14.0
pyparsing==3.0.9
python-dateutil==2.8.2
pytz==2022.7.1
PyYAML==6.0
requests==2.28.2
rich==13.3.2
six==1.16.0
sympy==1.11.1
tabulate==0.9.0
termcolor==2.2.0
torch==2.0.0
torchvision==0.15.1
typing_extensions==4.5.0
urllib3==1.26.15
yapf==0.32.0

 mmyolo git:(main) ✗ conda list
# packages in environment at /Users/pmctaggart/opt/miniconda3:
#
# Name                    Version                   Build  Channel
blas                      1.0                    openblas
brotlipy                  0.7.0           py39h1a28f6b_1002
bzip2                     1.0.8                h620ffc9_4
ca-certificates           2023.01.10           hca03da5_0
certifi                   2022.12.7        py39hca03da5_0
cffi                      1.15.1           py39h80987f9_3
charset-normalizer        2.0.4              pyhd3eb1b0_0
conda                     23.1.0           py39hca03da5_0
conda-package-handling    1.9.0            py39h1a28f6b_1
cryptography              38.0.1           py39h834c97f_0
ffmpeg                    4.2.2                h04105a8_0
filelock                  3.9.0            py39hca03da5_0
flit-core                 3.6.0              pyhd3eb1b0_0
freetype                  2.12.1               h1192e45_0
gettext                   0.21.0               h826f4ad_0
giflib                    5.2.1                h80987f9_3
gmp                       6.2.1                hc377ac9_3
gmpy2                     2.1.2            py39h8c48613_0
gnutls                    3.6.15               h887c41c_0
icu                       68.1                 hc377ac9_0
idna                      3.4              py39hca03da5_0
jinja2                    3.1.2            py39hca03da5_0
jpeg                      9e                   h80987f9_1
lame                      3.100                h1a28f6b_0
lcms2                     2.12                 hba8e193_0
lerc                      3.0                  hc377ac9_0
libcxx                    14.0.6               h848a8c0_0
libdeflate                1.8                  h1a28f6b_5
libffi                    3.4.2                hca03da5_6
libgfortran               5.0.0           11_3_0_hca03da5_28
libgfortran5              11.3.0              h009349e_28
libiconv                  1.16                 h1a28f6b_2
libidn2                   2.3.1                h1a28f6b_0
libopenblas               0.3.21               h269037a_0
libopus                   1.3                  h1a28f6b_1
libpng                    1.6.39               h80987f9_0
libtasn1                  4.16.0               h1a28f6b_0
libtiff                   4.5.0                h2fd578a_0
libunistring              0.9.10               h1a28f6b_0
libvpx                    1.10.0               hc377ac9_0
libwebp                   1.2.4                ha3663a8_1
libwebp-base              1.2.4                h80987f9_1
libxml2                   2.9.14               h8c5e841_0
llvm-openmp               14.0.6               hc6e5704_0
lz4-c                     1.9.4                h313beb8_0
markupsafe                2.1.1            py39h1a28f6b_0
mpc                       1.1.0                h8c48613_1
mpfr                      4.0.2                h695f6f0_1
mpmath                    1.2.1            py39hca03da5_0
ncurses                   6.3                  h1a28f6b_3
nettle                    3.7.3                h84b5d62_1
networkx                  2.8.4            py39hca03da5_0
numpy                     1.23.5           py39h1398885_0
numpy-base                1.23.5           py39h90707a3_0
openh264                  1.8.0                h98b2900_0
openssl                   1.1.1t               h1a28f6b_0
pillow                    9.4.0            py39h313beb8_0
pluggy                    1.0.0            py39hca03da5_1
pycosat                   0.6.4            py39h1a28f6b_0
pycparser                 2.21               pyhd3eb1b0_0
pyopenssl                 22.0.0             pyhd3eb1b0_0
pysocks                   1.7.1            py39hca03da5_0
python                    3.9.12               hbdb9e5c_0
python.app                3                py39h1a28f6b_0
pytorch                   2.0.0                   py3.9_0    pytorch
readline                  8.2                  h1a28f6b_0
requests                  2.28.1           py39hca03da5_0
ruamel.yaml               0.17.21          py39h1a28f6b_0
ruamel.yaml.clib          0.2.6            py39h1a28f6b_1
setuptools                65.5.0           py39hca03da5_0
sqlite                    3.40.0               h7a7dc30_0
sympy                     1.11.1           py39hca03da5_0
tk                        8.6.12               hb8d0fd4_0
toolz                     0.12.0           py39hca03da5_0
torchvision               0.15.0                 py39_cpu    pytorch
tqdm                      4.64.1           py39hca03da5_0
typing_extensions         4.4.0            py39hca03da5_0
tzdata                    2022g                h04d1e81_0
urllib3                   1.26.13          py39hca03da5_0
x264                      1!152.20180806       h1a28f6b_0
xz                        5.2.8                h80987f9_0
zlib                      1.2.13               h5a0b063_0
zstd                      1.5.2                h8574219_0
```

3. You may add addition that may be helpful for locating the problem, such as
   - How you installed PyTorch : conda
   - Other environment variables that may be related (such as `$PATH`, `$LD_LIBRARY_PATH`, `$PYTHONPATH`, etc.)

**Error traceback**
If applicable, paste the error trackback here.

```none
mmyolo git:(main) ✗ mim install "mmdet>=3.0.0rc6,<3.1.0"

Looking in links: https://download.openmmlab.com/mmcv/dist/cpu/torch2.0.0/index.html
Collecting mmdet<3.1.0,>=3.0.0rc6
  Using cached mmdet-3.0.0rc6-py3-none-any.whl (1.7 MB)
ERROR: Exception:
Traceback (most recent call last):
  File "/Users/pdmct/.local/share/virtualenvs/mmyolo-dGEsbyu_/lib/python3.11/site-packages/pip/_internal/cli/base_command.py", line 160, in exc_logging_wrapper
    status = run_func(*args)
             ^^^^^^^^^^^^^^^
  File "/Users/pdmct/.local/share/virtualenvs/mmyolo-dGEsbyu_/lib/python3.11/site-packages/pip/_internal/cli/req_command.py", line 247, in wrapper
    return func(self, options, args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pdmct/.local/share/virtualenvs/mmyolo-dGEsbyu_/lib/python3.11/site-packages/pip/_internal/commands/install.py", line 419, in run
    requirement_set = resolver.resolve(
                      ^^^^^^^^^^^^^^^^^
  File "/Users/pdmct/.local/share/virtualenvs/mmyolo-dGEsbyu_/lib/python3.11/site-packages/pip/_internal/resolution/resolvelib/resolver.py", line 92, in resolve
    result = self._result = resolver.resolve(
                            ^^^^^^^^^^^^^^^^^
  File "/Users/pdmct/.local/share/virtualenvs/mmyolo-dGEsbyu_/lib/python3.11/site-packages/pip/_vendor/resolvelib/resolvers.py", line 481, in resolve
    state = resolution.resolve(requirements, max_rounds=max_rounds)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pdmct/.local/share/virtualenvs/mmyolo-dGEsbyu_/lib/python3.11/site-packages/pip/_vendor/resolvelib/resolvers.py", line 373, in resolve
    failure_causes = self._attempt_to_pin_criterion(name)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pdmct/.local/share/virtualenvs/mmyolo-dGEsbyu_/lib/python3.11/site-packages/pip/_vendor/resolvelib/resolvers.py", line 213, in _attempt_to_pin_criterion
    criteria = self._get_updated_criteria(candidate)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pdmct/.local/share/virtualenvs/mmyolo-dGEsbyu_/lib/python3.11/site-packages/pip/_vendor/resolvelib/resolvers.py", line 203, in _get_updated_criteria
    for requirement in self._p.get_dependencies(candidate=candidate):
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pdmct/.local/share/virtualenvs/mmyolo-dGEsbyu_/lib/python3.11/site-packages/pip/_internal/resolution/resolvelib/provider.py", line 237, in get_dependencies
    return [r for r in candidate.iter_dependencies(with_requires) if r is not None]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pdmct/.local/share/virtualenvs/mmyolo-dGEsbyu_/lib/python3.11/site-packages/pip/_internal/resolution/resolvelib/provider.py", line 237, in <listcomp>
    return [r for r in candidate.iter_dependencies(with_requires) if r is not None]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pdmct/.local/share/virtualenvs/mmyolo-dGEsbyu_/lib/python3.11/site-packages/pip/_internal/resolution/resolvelib/candidates.py", line 247, in iter_dependencies
    requires = self.dist.iter_dependencies() if with_requires else ()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pdmct/.local/share/virtualenvs/mmyolo-dGEsbyu_/lib/python3.11/site-packages/mim/commands/install.py", line 247, in patched_iter_dependencies
    if 'mim' in self.iter_provided_extras:
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: argument of type 'method' is not iterable
```

**Bug fix**
If you have already identified the reason, you can provide the information here. If you are willing to create a PR to fix it, please also leave a comment here and that would be much appreciated!

**Comment by pdmct** [comment]
It was an issue with mismatched python versions.

**Comment by lyn12233** [comment]
It is a bug of missing parenthesis and is fixed by [#213 ](https://github.com/open-mmlab/mim/pull/213). However this fix is in v3.8.0 while mim installed by pip/conda seems to be v3.7.0. Recommend to install mim from source.
